janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java

```diff
@@ -504,6 +504,10 @@ private InternalVertex getExistingVertex(long vertexId) {
     }
 
     public InternalVertex getInternalVertex(long vertexId) {
+        if (vertexCache == null) {
```

Comment:
The class `StandardJanusGraphTx` already has a `verifyOpen()` method, which checks `isClosed()` and throws an exception if the transaction is closed. I quickly skimmed the `StandardJanusGraphTx` class and it looks to me like `isClosed()` is always equal to `vertexCache == null`, `addedRelations == null` and `deletedRelations == null`, but it's way more readable.

I suggest to add a `warnIfClosed()` method and call it here and also in `getExistingVertex()`, `getInternalVertices()`, `connectRelation()`, `getSchemaVertex()` and `expireSchemaElement()`.

Reply:
I don't know exactly how transactions are handled between threads, but if multiple threads can access the same transaction, we need to synchronize state-changing calls to the same transaction.

Reply:
> if multiple threads can access the same transaction, we need to synchronize state-changing calls to the same transaction

TBH I am not sure if race condition is the cause for the NPE in `getInternalVertex`.

Reply:
Me neither and it's hard to debug as well

Reply:
I now assign a dummy empty cache to the variable `vertexCache` in transaction release method.